subprojects/build-operations/src/main/java/org/gradle/trace/buildops/BuildOperationTrace.java

```diff
@@ -44,15 +41,11 @@ public BuildOperationTrace(GradleInternal gradle) {
     }
 
     public BuildOperationTrace measureGarbageCollection(File outFile) {
-        gradle.buildFinished(result -> {
-            try (PrintWriter writer = new PrintWriter(outFile)) {
-                writer.println(ManagementFactory.getGarbageCollectorMXBeans().stream()
-                    .map(GarbageCollectorMXBean::getCollectionTime)
-                    .reduce(0L, Long::sum));
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
-            }
+        Provider<GcTimeCollectionService> provider = sharedServices.registerIfAbsent("gc-time", GcTimeCollectionService.class, spec -> {
+            spec.getParameters().getOutputFile().set(outFile);
         });
+        // Force the service to be instantiated so we actually get a close() call at the end of the build
+        provider.get();
```

Comment:
I don't think that's what happens exactly, but there is indeed a problem. I extended the test to check if the GC time is non-zero. And indeed with CC it _is_ zero...

Reply:
The build service won't be restarted from the cache if it's not actually required by any tasks, providers or as a listener.

In this case it would make sense to register it as a build events listener with `BuildEventsListenerRegistry`.

Reply:
That is the trick employed by the other listeners here. But it means that I'll get a callback for potentially hundreds of thousands of events, which adds a significant overhead (=bias) to the benchmark, so I'd like to avoid doing so.

Are there any other ways to solve this without getting a ton of unnecessary callbacks?

Reply:
Maybe you can inject a task with the sole purpose of initializing the service?

Reply:
It's fixed now using a fake `OperationCompletionListener`.